### ISSUE TITLE
all: static analysis to proactively free constructed garbage

### ISSUE THREAD

**Issue by bradfitz** [golang/go#5448]
<pre>In the following program, the compiler+linker could know that NewIntWriter always
returns unretained garbage and that on exit from foo, no reference is held to
*IntWriter, and an explicit free could be inserted.

You see this with things like <a href="http://golang.org/pkg/encoding/json/#NewEncoder">http://golang.org/pkg/encoding/json/#NewEncoder</a> (see also:
<a href="https://golang.org/cl/9365044/)">https://golang.org/cl/9365044/)</a>

Instead, it generates garbage:

BenchmarkAllocs  2000000           819 ns/op        1057 B/op          1 allocs/op

package main

import (
        &quot;fmt&quot;
        &quot;io&quot;
        &quot;io/ioutil&quot;
        &quot;testing&quot;
)

type IntWriter struct{
        w   io.Writer
        state [1000]byte
}

func NewIntWriter(w io.Writer) *IntWriter {
        return &amp;IntWriter{w: w}
}

func (iw *IntWriter) WriteInt(n int) {
        fmt.Fprintf(iw.w, &quot;%d&quot;, n)
}

func BenchmarkAllocs(b *testing.B) {
        b.ReportAllocs()
        for i := 0; i &lt; b.N; i++ {
                foo()
        }
}

func foo() {
        NewIntWriter(ioutil.Discard).WriteInt(123)
}</pre>

**Comment by dvyukov** [comment]
It is already 2 allocs:

BenchmarkAllocs	 2000000	       890 ns/op	    1040 B/op	       2 allocs/op

Looks like a target for escape analysis.

**Comment by pwaller** [comment]
I believe that escape analysis already covers this today:

```
BenchmarkAllocs-16    	20000000	        85.6 ns/op	       8 B/op	       1 allocs/op
```

(also, wow things got faster in the interim - is that really 10x faster than the above result?)

For the record, here is the salient information:

```
(pprof) top
Showing nodes accounting for 9699476, 100% of 9699476 total
      flat  flat%   sum%        cum   cum%
   9699476   100%   100%    9699476   100%  github.com/pwaller/escape-analysis.(*IntWriter).WriteInt
         0     0%   100%    9699476   100%  github.com/pwaller/escape-analysis.BenchmarkAllocs
         0     0%   100%    9699476   100%  github.com/pwaller/escape-analysis.foo
         0     0%   100%    9699476   100%  testing.(*B).launch
         0     0%   100%    9699476   100%  testing.(*B).runN
(pprof) list WriteInt
Total: 9699476
ROUTINE ======================== github.com/pwaller/escape-analysis.(*IntWriter).WriteInt in /home/pwaller/.local/src/github.com/pwaller/escape-analysis/main_test.go
   9699476    9699476 (flat, cum)   100% of Total
         .          .     15:func NewIntWriter(w io.Writer) *IntWriter {
         .          .     16:        return &IntWriter{w: w}
         .          .     17:}
         .          .     18:
         .          .     19:func (iw *IntWriter) WriteInt(n int) {
   9699476    9699476     20:        fmt.Fprintf(iw.w, "%d", n)
         .          .     21:}
         .          .     22:
         .          .     23:func BenchmarkAllocs(b *testing.B) {
         .          .     24:        b.ReportAllocs()
         .          .     25:        for i := 0; i < b.N; i++ {
```

```
$ go build -gcflags=-m
# github.com/pwaller/escape-analysis
./main.go:15:6: can inline NewIntWriter
./main.go:19:6: can inline (*IntWriter).WriteInt
./main.go:30:6: can inline foo
./main.go:31:21: inlining call to NewIntWriter
./main.go:31:46: inlining call to (*IntWriter).WriteInt
./main.go:24:23: inlining call to testing.(*B).ReportAllocs
./main.go:26:20: inlining call to foo
./main.go:26:20: inlining call to NewIntWriter
./main.go:26:20: inlining call to (*IntWriter).WriteInt
/tmp/go-build401243069/b001/_gomod_.go:5:6: can inline init.0
./main.go:16:27: &IntWriter literal escapes to heap
./main.go:15:19: leaking param: w to result ~r1 level=-1
./main.go:19:7: leaking param content: iw
./main.go:20:27: n escapes to heap
./main.go:20:20: (*IntWriter).WriteInt ... argument does not escape
./main.go:26:20: n escapes to heap
./main.go:23:22: BenchmarkAllocs b does not escape
./main.go:26:20: BenchmarkAllocs &IntWriter literal does not escape
./main.go:26:20: BenchmarkAllocs ... argument does not escape
./main.go:31:46: n escapes to heap
./main.go:31:21: foo &IntWriter literal does not escape
./main.go:31:46: foo ... argument does not escape
# github.com/pwaller/escape-analysis
```

Seems the remaining allocation is the interface argument to `fmt.Fprintf`. Slightly surprised that requires an allocation - I thought values which could fit inside the interface wouldn't get allocated on the heap. Unless I'm misunderstanding where the allocation is here.

/cc @mvdan for gardening on the issue, since I believe the issue as originally written is resolved and I don't have rights to close the issue.

**Comment by mvdan** [comment]
> (also, wow things got faster in the interim - is that really 10x faster than the above result?)

I presume your machine is just more powerful :)

I agree that it does seem like the issue as written has been fixed. There already are other open issues to improve the compiler, so we can continue work elsewhere.

**Comment by pwaller** [comment]
TIL from [a conversation on #performance on gophers slack](https://gophers.slack.com/archives/C0VP8EF3R/p1549401040972800) that interface data always contains pointer, so there has to be an allocation in these circumstances, currently.

Doc trail in case anyone find themselves here:

* https://commaok.xyz/post/interface-allocs/
* (closed) [#8405: cmd/gc: make interface updates atomic wrt garbage collector](https://github.com/golang/go/issues/8405)
* (open) [#8618: runtime: don't allocate for non-escaping conversions to interface{}](#8618)
